### ISSUE TITLE
Reinvoke shall never be invoked again ;-)

### ISSUE THREAD

**Issue by bgrundmann** [TriOptima/iommi#92]
This is a proof of concept sketch of an alternative implementation of iommi's system of styling + setting / collecting members.

It doesn't do a bunch of things the real thing does (shortcuts, styles from the super class of a class, position arguments to __init__).  But I think all of them should be reasonably straightforward to do.

This is all just to kickstart a conversation ;-)

```python
from tri_declarative import Namespace, dispatch

registered_styles = {
    'Foo': dict(
        points__stylish_point__x=3,
        points__stylish_point__y=5
    ),
    'Line': dict(
        a__x=7,
        b__x=17,
        b__y=20
    )
}


def iommi(members={}, member_namespaces={}):
    """A class which can be constructed and styled the iommi way.
       In particular it can have named members and named namespaces of members
       which can again be styled based on their class name and implicitely
       constructed by just naming one of their attributes.

       Unlike the current iommi implementation this avoids calling the corresponding
       classes __init__ more than once (the infamous reinvoke).  Instead it replaces
       the __init__ by a trival one that just stores the argument for later.

       Rationale: We only want to call the real init once we know what our
       containing classes (aka iommi_parents) are -- and in particular the
       arguments they are adding to our init using dunder_paths.

       This also has the advantage that the real init is called at the same time
       and the same number of times for classes that are implicitely constructed
       and those that are explicitely constructed.

       So in a change of API the outermost object (e.g. iommi_root) would have
       to call iommi_finish_init(Namespace()).

       That then will trigger recursive calls of iommi_finish_init, looking up the
       style information and eventually calling the original init.
    """
    def decorator(cls):
        original_init = cls.__init__

        def set_member(namespace, name, constructor, style_from_parent, none_or_config_or_value):
            """Set namespace[name] = value. if necessary construct the value first.
            """
            assert(name not in namespace)
            if none_or_config_or_value is None:
                pass
            else:
                # At this point we have all the arguments.  So we can initialize the value
                value = none_or_config_or_value
                if isinstance(none_or_config_or_value, dict):
                    value = constructor(**none_or_config_or_value)
                namespace[name] = value
                if hasattr(value, 'iommi_finish_init'):
                    value.iommi_finish_init(style_from_parent)

        @dispatch
        def __init__(self, *args, **kwargs):
            # TODO: actually support *args
            self._iommi_init_kwargs = kwargs

        def iommi_finish_init(self, style_from_parent):
            style_kwargs = Namespace(style_from_parent, registered_styles.get(cls.__name__, {}))
            kwargs = Namespace(style_kwargs, self._iommi_init_kwargs)
            # To avoid holding onto more memory then necessary
            for name, constructor in members.items():
                config = kwargs.pop(name, None)
                set_member(kwargs, name, constructor, style_kwargs.get(name, {}), config)
            for namespace_name, constructor in member_namespaces.items():
                configs = kwargs.pop(namespace_name, {})
                if configs is None:
                    continue
                member_namespace = Namespace()
                kwargs[namespace_name] = member_namespace
                for name, config in configs.items():
                    set_member(member_namespace, name, constructor, style_kwargs.get(namespace_name, {}).get(name, {}), config)
            original_init(self, **kwargs)
            del self._iommi_init_kwargs

        cls.iommi_finish_init = iommi_finish_init
        cls.__init__ = __init__
        return cls

    return decorator


@iommi()
class Point:
    @dispatch
    def __init__(self, x=-5, y=-5):
        self.x = x
        self.y = y

    def __str__(self):
        return f"{self.x}x{self.y}"


@iommi(members=dict(a=Point, b=Point))
class Line:
    def __init__(self, a, b):
        self.a = a
        self.b = b

    def __str__(self):
        return f"{self.a}-{self.b}"


@iommi(members=dict(origin=Point), member_namespaces=dict(points=Point, lines=Line))
class Foo:
    def __init__(self, origin, points, lines, **kwargs):
        self.origin = origin
        self.points = points
        self.lines = lines
        self.kwargs = kwargs

    def __str__(self):
        return f"origin={self.origin} points={self.points} lines={self.lines} kwargs={self.kwargs}"


instance = Foo(origin__x=1,
               origin__y=2,
               points__a__x=24,
               points__a__y=42,
               points__b=Point(x=100, y=100),
               # Uncomment the below to make stylish point disappear.
               # points__stylish_point=None,
               lines__one__a=Point(x=0, y=0),
               lines__one__b__y=10)


instance.iommi_finish_init(Namespace())

print(instance)
```

**Comment by boxed** [comment]
We've also discussed ways to get rid of `reinvoke`. The basic idea you have and Johan has seem similar, even if the implementation is very different.

We were talking about maybe introducing an explicit third step between `__init__` and `bind` to handle some of this. But we do still want the validation we do now when constructing an object to happen, because it is a big reason we crash early on errors instead of late.

**Comment by bgrundmann** [comment]
I *think* my implementation does whatever validation there is as soon as it
makes sense — that is as soon as we know that we have all parameters...

**Comment by boxed** [comment]
Well that might actually be too late for my personal taste. Bind-time is too late for example. We do lots of validation at declaration time which makes it harder to write broken code.

**Comment by bgrundmann** [comment]
Hmm we might be talking past each other, but: 

- My proposed implementation doesn't do it at bind time but earlier.

- Validation of not actually all the parameters seems - eh - silly.  I mean I can clearly construct examples where a Style or a Meta parameter of a parent will make a call to a sub components init fail (or vice versa it will fail until those are supplied). 

If you want we can do another zoom session and try to hash it out quickly.

**Comment by boxed** [comment]
> My proposed implementation doesn't do it at bind time but earlier.

Sure, but not at declaration time right? 

> Validation of not actually all the parameters seems - eh - silly.

Hmm.. you might be right. It IS annoying when all that code is run when you want to debug a specific test, just because some stuff is declared globally :P

> If you want we can do another zoom session and try to hash it out quickly.

I'll talk with Johan about it. I think he's at lunch right now :P

**Comment by bgrundmann** [comment]
To be clear I'm not wedded to the proposed implementation, but I think the proposed semantics is a little less crazy than the current one.  (Still crazy though it is not what people would expect would happen to their init function).  But on the pro side:

- Order of init functions is right (that is the original init functions of sub components are called first)
- Number of calls to each init function is what you would expect (e.g. once)

**Comment by bgrundmann** [comment]
And lastly I'm using words like crazy -- but I very much like what iommi offers and have not seen it before, so no hard feelings... ;-)

**Comment by boxed** [comment]
I think we all three agree that lots of things in iommi are a bit crazy :P

**Comment by jlubcke** [comment]
I added a testcase (`test_get_config`) to have a single place showing all the different ways config propagate. I think it is good to have a baseline when we start to fiddle with fundamental parts...

**Comment by boxed** [comment]
Fixed in iommi 4.0!
